### PR TITLE
WT-6743 Save and restore session's data handle during cursor reopen call

### PR DESCRIPTION
**Pull request by ddanderson**
Since the internal reopen call may be called on cursors unrelated to the current API call (as part of cursor cache sweeping), it needs to save and restore the session's data handle.  This change also includes extra assertions.

Setting statistics for reopen now happens in the "file cursor" code, as that is where the session data handle is valid.

### CODE REVIEW

**Comment by ddanderson** [comment]
@sueloverso and @keitharnoldsmith, can you both please review and see if this makes sense?  I realize it's late on Friday, there's no rush on this.  The summary of the pull request has a blurb of the changes (and will be the commit comment).  Thanks!

File: src/cursor/cur_file.c [hunk 1]
```diff
@@ -600,11 +593,47 @@ __curfile_reopen(WT_CURSOR *cursor, bool check_only)
...
+         * happen if any session (including ours) is actively using it.
+         */
+        WT_ASSERT(session, can_reopen || dhandle != session->dhandle);
+        return (can_reopen ? 0 : WT_NOTFOUND);
     }
+
+    /*
+     * Temporarily set the session's data handle. Reopen may be called either as part of an open API
+     * call, or during cursor sweep as part of a different API call, so we need to restore the data
+     * handle.
```
Comment (sueloverso) [inline src/cursor/cur_file.c:634]: Minor clarification on the comment. It seems like the first sentence and the last couple of words refer to different data handles? I'm not entirely sure. Maybe something like this (if it is correct):

Temporarily set the session's data handle to the dhande in this cursor. Reopen may be called... different API call, so we need to restore the original data handle that was in our session after reopen completes.

**Review by sueloverso — APPROVED** [review]
LGTM. I have one question on the comment.

File: src/cursor/cur_file.c (hunk 1 above)
Comment (ddanderson) [reply src/cursor/cur_file.c]: Thanks, you're right.  I updated the comment to your wording.

**Review by keitharnoldsmith — APPROVED** [review]
LGTM!

**Pull request merged**